File: backend/src/routes/legal.ts
import { Router, type Response } from 'express'
import { requireAuth, type AuthRequest } from '../middleware/auth.js'
import { acceptCurrentLegalForUser, getLegalStatusForUser } from '../lib/legalDocuments.js'

import { ErrorCodes, sendError } from '../lib/apiErrors.js'
import { asyncHandler } from '../lib/httpErrors.js'
import { getLegalDocument } from '../services/legalService.js'

const router = Router()

function slugParamToEnum(slug: string): 'TERMS' | 'PRIVACY' | null {
  if (slug === 'terms') return 'TERMS'
  if (slug === 'privacy') return 'PRIVACY'
  return null
}

// GET /api/legal/status/me
router.get(
  '/status/me',
  requireAuth,
  asyncHandler(async (req: AuthRequest, res: Response) => {
    const status = await getLegalStatusForUser(req.userId!)
    if (!status) {
      sendError(res, 404, ErrorCodes.USER_NOT_FOUND)
      return
    }
    res.json(status)
  })
)

// POST /api/legal/accept
router.post(
  '/accept',
  requireAuth,
  asyncHandler(async (req: AuthRequest, res: Response) => {
    const accepted = await acceptCurrentLegalForUser(req.userId!)
    res.json({ ok: true, ...accepted })
  })
)

// GET /api/legal/terms | /api/legal/privacy
router.get(
  '/:slug',
  asyncHandler(async (req, res: Response) => {
    const slug = slugParamToEnum(String(req.params.slug ?? '').toLowerCase())
    if (!slug) {
      sendError(res, 404, ErrorCodes.LEGAL_DOCUMENT_NOT_FOUND)
      return
    }
    const rawLocale =
      typeof req.query.locale === 'string'
        ? req.query.locale
        : req.headers['accept-language']?.split(',')[0]
    res.json(await getLegalDocument(slug, rawLocale))
  })
)

export default router

File: backend-rust/services/api-gateway/src/handlers/data/legal.rs
use axum::{
    Json,
    extract::{Query, State},
    http::{HeaderMap, header},
};
use serde::Deserialize;

use crate::AppState;
use crate::handlers::auth::routes::api_error_response;
use crate::middleware::auth::AuthUser;

#[derive(Debug, Deserialize)]
pub struct LegalLocaleQuery {
    pub locale: Option<String>,
}

pub async fn legal_status_handler(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<
    Json<streakmeet_legal::LegalStatusResponse>,
    (axum::http::StatusCode, Json<serde_json::Value>),
> {
    match streakmeet_legal::get_legal_status_for_user(&state.pool, &auth.user_id).await {
        Ok(Some(status)) => Ok(Json(status)),
        Ok(None) => Err(api_error_response(streakmeet_types::ApiError::new(
            404,
            streakmeet_types::codes::USER_NOT_FOUND,
            None,
        ))),
        Err(err) => Err(api_error_response(err)),
    }
}

pub async fn legal_accept_handler(
    State(state): State<AppState>,
    auth: AuthUser,
) -> Result<Json<serde_json::Value>, (axum::http::StatusCode, Json<serde_json::Value>)> {
    let accepted = streakmeet_legal::accept_current_legal_for_user(&state.pool, &auth.user_id)
        .await
        .map_err(api_error_response)?;
    Ok(Json(serde_json::json!({
        "ok": true,
        "terms": accepted.terms,
        "privacy": accepted.privacy,
    })))
}

pub async fn legal_document_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(query): Query<LegalLocaleQuery>,
    axum::extract::Path(slug): axum::extract::Path<String>,
) -> Result<
    Json<streakmeet_legal::LegalDocumentJson>,
    (axum::http::StatusCode, Json<serde_json::Value>),
> {
    let Some(legal_slug) = streakmeet_legal::LegalSlug::from_param(&slug) else {
        return Err(api_error_response(streakmeet_types::ApiError::new(
            404,
            streakmeet_types::codes::LEGAL_DOCUMENT_NOT_FOUND,
            None,
        )));
    };

    let raw_locale = query.locale.as_deref().or_else(|| {
        headers
            .get(header::ACCEPT_LANGUAGE)
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.split(',').next())
    });

    streakmeet_legal::get_legal_document(&state.pool, legal_slug, raw_locale)
        .await
        .map(Json)
        .map_err(api_error_response)
}
